ring/ring_poly.go

```diff
@@ -3,36 +3,49 @@ package ring
 import (
 	"encoding/binary"
 	"errors"
-	"math/bits"
-
-	"github.com/tuneinsight/lattigo/v3/utils"
 )
 
 // Poly is the structure that contains the coefficients of a polynomial.
 type Poly struct {
-	Coeffs  [][]uint64 // Coefficients in CRT representation
+	Coeffs  [][]uint64 // Dimension-2 slice of coefficients (re-slice of Buff)
+	Buff    []uint64   // Dimension-1 slice of coefficient
 	IsNTT   bool
 	IsMForm bool
 }
 
-// NewPoly creates a new polynomial with N coefficients set to zero and nbModuli moduli.
-func NewPoly(N, nbModuli int) (pol *Poly) {
+// NewPoly creates a new polynomial with N coefficients set to zero and Level+1 moduli.
+func NewPoly(N, Level int) (pol *Poly) {
 	pol = new(Poly)
-	pol.Coeffs = make([][]uint64, nbModuli)
-	for i := 0; i < nbModuli; i++ {
-		pol.Coeffs[i] = make([]uint64, N)
+
+	pol.Buff = make([]uint64, N*(Level+1))
+	pol.Coeffs = make([][]uint64, Level+1)
+	for i := 0; i < Level+1; i++ {
+		pol.Coeffs[i] = pol.Buff[i*N : (i+1)*N]
 	}
+
 	return
 }
 
-// Degree returns the number of coefficients of the polynomial, which equals the degree of the Ring cyclotomic polynomial.
-func (pol *Poly) Degree() int {
-	return len(pol.Coeffs[0])
+// Resize resize the level of the target polynomial to the provided level.
+// If the provided level is larger than the current level, then allocates zero
+// coefficients, else dereferences the coefficients above the provided level.
+func (pol *Poly) Resize(level int) {
```

Comment:
Allow down sizing but not up sizing, since up sizing is equivalent to re-allocation.

Reply:
I would actually keep it that way. This is a low level method and it cannot be abused in the schemes to increase the levels without inducing correctness errors.

Reply:
I don't like that the evaluator does resizing under the hood as the user might not expect that its ciphertext's internal buffer change size. But since it only resizes output ciphertexts and since this behaviour can be changed with the same interface, we can move ahead with it.